Adding commandline options to list boards and ports

makenai opened the pull request:
Closes #1 

Added commandline options to list boards and ports. Started adding some test boilerplate to build upon and tests for board detection.

Let me know if any of the output or refactoring isn't up to spec and I can fix it up.

noopkat commented:
@makenai thank you for this! These are really nice ideas/additions. I have some feedback for this PR, but need a little time to write it up. Stay tuned!

noopkat commented on README.md:
````diff
@@ -73,7 +73,30 @@ var avrgirl = new Avrgirl({
+You can list available USB ports programmatically using the the `listPorts` method:
+
+```javascript
+Avrgurl.listPorts(function(err, ports) {
````
spelling of Avrgirl

makenai replied:
Doh!

noopkat commented on README.md:
````diff
@@ -73,7 +73,30 @@ var avrgirl = new Avrgirl({
+```
+$ avrgirl-arduino list
+Listing Devices:
+ - port: /dev/cu.usbmodem1421 / pid: 0x0043 / mfg: Arduino (www.arduino.cc) / board: uno
````
the forward slashes make this look a little confusing and heard to read IMO. This output also wouldn't pipe super well into anything other than grep. Would perhaps a more minimal JSON object to the raw one suffice instead?

noopkat commented on README.md:
```diff
@@ -110,12 +133,21 @@ Required flags:
+
+`avrgirl-arduino list [--raw]`
+
+If you add the `--raw` flag, the output will be in a very verbose JSON format.
```
I like this idea a lot

noopkat commented on avrgirl-arduino.js:
```diff
@@ -129,6 +129,8 @@ Avrgirl_arduino.prototype._upload = function (hex, callback) {
     });
   }
+  // this is a success callback
+  callback();
```
The callbacks will occur for this method within each protocol's upload call. A callback at the bottom of this method could potentially cause a false (and early) success. Perhaps if it doesn't find the correct protocol, a callback is returned with an error instead? Within an `else` at the end of the list of conditionals.

Let me know if I'm missing something here or the above suggestion is not the solution for what you were trying to solve.

makenai replied:
Oops, that's exactly right. I'll fix this up.

noopkat commented on bin/cli.js:
```diff
@@ -20,30 +24,51 @@ handleInput(userAction, argv);
+            var port = ports[i];
+            if ( port._standardPid ) {
+              console.log(
+                util.format(" - port: %s / pid: %s / mfg: %s / board: %s",
```
see earlier comment on the documentation summary of this

noopkat commented on bin/cli.js:
```diff
@@ -20,30 +24,51 @@ handleInput(userAction, argv);
+      });
       break;
     case 'help':
-      showHelp();
```
I'm curious about the reason for removing what's inside the help case here

makenai replied:
These seemed like redundant calls to showHelp(). Since 

```
case 'help':
case undefined:
default:
```

all take the same code path, we can take out the breaks and have them fall through to the same call to showHelp. If this isn't your style, I can revert it.

noopkat replied:
The reasoning for keeping things separate was to exit with the correct status code. I feel the case switch format keeps this clear and readable so you can see what input exits with an error.

noopkat commented on bin/cli.js:
```diff
@@ -20,30 +24,51 @@ handleInput(userAction, argv);
-      showHelp();
-      return process.exit();
     case undefined:
-      showHelp();
```
see previous comment above about the help case :grin:

noopkat replied:
I'm ok with merging `undefined` and `default`

noopkat commented on boards.js:
```diff
@@ -1,4 +1,4 @@
-module.exports = {
+var boards = {
```
this is all a cool improvement to the boards file

noopkat commented on tests/board-detection.spec.js:
```diff
@@ -0,0 +1,55 @@
+  var avrgirl = new Avrgirl({
+    board: 'uno',
+  });
+  t.notOk(avrgirl.options.port, 'no port was passed in');
```
does this particular assert look confusing when you run the tests?
Ie. Does it show up as:

:white_check_mark: no port was passed in

and is it a check to make sure that the test setup was created properly, or is it asserting something about avrgirl-arduino itself when it instantiates?

makenai replied:
I think it's kind of a mix of the two and probably not needed. I want to make sure that the initial state is that no port was specified (in the setup) or defaulted to (by avrgirl), and that a port is autodetected when we flash. I could reword it was 'initialised without a port', but I think it's best to remove.

noopkat commented on tests/board-detection.spec.js:
```diff
@@ -0,0 +1,55 @@
+
+test('pid board detection', function(t) {
+  t.plan(2);
+  var Avrgirl = proxyquire('../avrgirl-arduino', {
```
to be safer from false failure reporting on Travis, this proxyquire call should also have `noCallThru` at the beginning of the chain. See an example here: https://github.com/noopkat/avrgirl-ispmkii/blob/master/tests/avrgirl-ispmkii.spec.js#L9

noopkat commented on tests/board-detection.spec.js:
```diff
@@ -0,0 +1,55 @@
+
+test('pnpId board detection', function(t) {
+  t.plan(2);
+  var Avrgirl = proxyquire('../avrgirl-arduino', {
```
to be safer from false failure reporting on Travis, this proxyquire call should also have `noCallThru` at the beginning of the chain. See an example here: https://github.com/noopkat/avrgirl-ispmkii/blob/master/tests/avrgirl-ispmkii.spec.js#L9

makenai replied:
Will fix both of these.

noopkat commented on tests/helpers/mockSerial.js:
```diff
@@ -0,0 +1,47 @@
+/**
+ * Provides a somewhat complete mock implementation of node-serialport
```
this is a great mock, will be adapting for other libraries, thanks!

makenai commented:
Thanks for the feedback, Suz! Give me a few days and I'll address the issues.

makenai commented:
Sorry it took me a bit longer than I thought to get back to this. I've incorporated your suggested changes and rebased against master. How does this look?

makenai commented:
@noopkat Patch up to preserve the exit status - thanks for explaining!

noopkat commented:
Thanks @makenai! Merged and will publish soon with some other additions (like CI stuff and readme changes). 

Just a small note - for now I have consolidated the `list` CLI method to just output the verbose version only. The reasoning is that I'd prefer to speak to some active `avrgirl-arduino` users first and gather some requirements and ideas from them first before setting anything in stone for this. I'm sorry to give you the runaround for that JSON object in the end, I hope you understand.

I really appreciate these contributions though, so thanks again :clap:

makenai commented:
Thank you, Susan.
